ee/server/service/software_installers.go

```diff
@@ -116,6 +117,232 @@ func preProcessUninstallScript(payload *fleet.UploadSoftwareInstallerPayload) {
+			payload.StorageID = payloadForNewInstallerFile.StorageID
+			payload.Filename = payloadForNewInstallerFile.Filename
+			payload.Version = payloadForNewInstallerFile.Version
+			payload.PackageIDs = payloadForNewInstallerFile.PackageIDs
+
+			dirty = append(dirty, "Package")
+		} else { // noop if uploaded installer is identical to previous installer
+			payloadForNewInstallerFile = nil
+			payload.InstallerFile = nil
+		}
+	}
+
+	if payload.InstallerFile == nil { // fill in existing existingInstaller data to payload
+		payload.StorageID = existingInstaller.StorageID
+		payload.Filename = existingInstaller.Name
+		payload.Version = existingInstaller.Version
+		payload.PackageIDs = existingInstaller.PackageIDs()
+	}
+
+	// default pre-install query is blank, so blanking out the query doesn't have a semantic meaning we have to take care of
+	if payload.PreInstallQuery != nil && *payload.PreInstallQuery != existingInstaller.PreInstallQuery {
+		dirty = append(dirty, "PreInstallQuery")
+	}
+
+	if payload.InstallScript != nil {
+		installScript := file.Dos2UnixNewlines(*payload.InstallScript)
+		if installScript == "" {
+			installScript = file.GetInstallScript(existingInstaller.Extension)
+		}
+
+		if installScript != existingInstaller.InstallScript {
+			dirty = append(dirty, "InstallScript")
+			payload.InstallScript = &installScript
+		}
+	}
+
+	if payload.PostInstallScript != nil {
+		postInstallScript := file.Dos2UnixNewlines(*payload.PostInstallScript)
+		if postInstallScript != existingInstaller.PostInstallScript {
+			dirty = append(dirty, "PostInstallScript")
+			payload.PostInstallScript = &postInstallScript
+		}
+	}
+
+	if payload.UninstallScript != nil {
+		uninstallScript := file.Dos2UnixNewlines(*payload.UninstallScript)
+		if uninstallScript == "" { // extension can't change on an edit so we can generate off of the existing file
+			uninstallScript = file.GetUninstallScript(existingInstaller.Extension)
+		}
+
+		payloadForUninstallScript := &fleet.UploadSoftwareInstallerPayload{
+			Extension:       existingInstaller.Extension,
+			UninstallScript: uninstallScript,
+			PackageIDs:      existingInstaller.PackageIDs(),
+		}
+		if payloadForNewInstallerFile != nil {
+			payloadForUninstallScript.PackageIDs = payloadForNewInstallerFile.PackageIDs
+		}
+
+		preProcessUninstallScript(payloadForUninstallScript)
+		if payloadForUninstallScript.UninstallScript != existingInstaller.UninstallScript {
+			uninstallScript = payloadForUninstallScript.UninstallScript
+			dirty = append(dirty, "UninstallScript")
+			payload.UninstallScript = &uninstallScript
+		}
+	}
+
+	// persist changes starting here, now that we've done all the validation/diffing we can
+	if len(dirty) > 0 {
+		if slices.Equal(dirty, []string{"SelfService"}) { // only self-service changed; use lighter update function
+			if err := svc.ds.UpdateInstallerSelfServiceFlag(ctx, *payload.SelfService, existingInstaller.InstallerID); err != nil {
```

Comment:
Let me see if I can add that. Will need to figure out whether that's doable cleanly across multiple data store calls, or whether I'll need to collapse thsi down into one call.

Reply:
Collapsed the side effects into a single call, since they're related. I think it's valid for an installer to be updated but side effects be rolled back if it comes to that, as long as _all_ side effects are rolled back (which as of a9fae1da37e8ab15c8aa44190fe223acf32e22c4 they are), so keeping that outside the transaction rather than trying to roll it in (and this also makes the side effects method reusable for GitOps).

Speaking of GitOps, that's why the signature for the new data store method is the way it is. You'll see that in an upcoming PR (not this one).